app/bundles/DynamicContentBundle/Helper/DynamicContentHelper.php

```diff
@@ -118,4 +137,110 @@ public function replaceTokensInContent($content, $lead)
 
         return $content;
     }
+
+    /**
+     * @param string    $content
+     * @param Lead|null $lead
+     *
+     * @return array
+     */
+    public function findDwcTokens($content, $lead)
+    {
+        preg_match_all('/{dwc=(.*?)}/', $content, $matches);
+
+        $tokens = [];
+        if (!empty($matches[1])) {
+            foreach ($matches[1] as $key => $slotName) {
+                $token = $matches[0][$key];
+                if (!empty($tokens[$token])) {
+                    continue;
+                }
+
+                $dwcs = $this->getDwcsBySlotName($slotName);
+
+                /** @var DynamicContent $dwc */
+                foreach ($dwcs as $dwc) {
+                    if ($dwc->getIsCampaignBased()) {
+                        continue;
+                    }
+                    $content                   = $lead ? $this->getRealDynamicContent($dwc->getSlotName(), $lead, $dwc) : '';
+                    $tokens[$token]['content'] = $content;
+                    $tokens[$token]['filters'] = $dwc->getFilters();
+                }
+            }
+
+            unset($matches);
+        }
+
+        return $tokens;
+    }
+
+    /**
+     * @param $slot
+     * @param $lead
+     * @param $dwc
+     *
+     * @return string
+     */
+    public function getRealDynamicContent($slot, $lead, DynamicContent $dwc)
+    {
+        $content = $dwc->getContent();
+        // Determine a translation based on contact's preferred locale
+        /** @var DynamicContent $translation */
+        list($ignore, $translation) = $this->dynamicContentModel->getTranslatedEntity($dwc, $lead);
+        if ($translation !== $dwc) {
+            // Use translated version of content
+            $dwc     = $translation;
+            $content = $dwc->getContent();
+        }
+        $this->dynamicContentModel->createStatEntry($dwc, $lead, $slot);
+
+        $tokenEvent = new TokenReplacementEvent($content, $lead, ['slot' => $slot, 'dynamic_content_id' => $dwc->getId()]);
+        $this->dispatcher->dispatch(DynamicContentEvents::TOKEN_REPLACEMENT, $tokenEvent);
+
+        return $tokenEvent->getContent();
+    }
+
+    /**
+     * @param string $slotName
+     *
+     * @return array|\Doctrine\ORM\Tools\Pagination\Paginator
+     */
+    public function getDwcsBySlotName($slotName)
+    {
+        return $this->dynamicContentModel->getEntities(
```

Comment:
It is alway better to encapsulate implementation. This should only call something like `return $this->dynamicContentModel->getEntitiesBySlotName($slotName)` and query should be builded inside of Repository class.